### PR TITLE
OnCall prometheus metrics exporter

### PR DESCRIPTION
**Pull request by Ferril**
# What this PR does
Add OnCall prometheus metrics exporter

## Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated

### CODE REVIEW

File: engine/settings/base.py
Comment (iskhakov) [inline engine/settings/base.py:219]: I know this is still in draft, adding this comment just to doublecheck that un-authenticated /metrics/ endpoint won't be exposed outside the cluster before merging

**Review by iskhakov — DISMISSED** [review]

File: engine/apps/metrics_exporter/helpers.py
Comment (Ferril) [reply engine/apps/metrics_exporter/helpers.py:107]: Yes, you are right, it should be "No team" now 👍

File: engine/apps/metrics_exporter/helpers.py
Comment (Ferril) [reply engine/apps/metrics_exporter/helpers.py:1]: Yes, each metric will be cached per organization

File: engine/apps/alerts/models/alert_group.py
Comment (Ferril) [reply engine/apps/alerts/models/alert_group.py:1113]: It just works faster for the same result in mysql

**Comment by Ferril** [comment]
> Have you considered calculating this on-the-fly during the background task which calculates the metrics? Something like this:
> 
> ```python
> response_time = alert_group.get_first_activity_log_timestamp() - alert_group.started_at
> ```

It's a good idea if we were talking only about the calculation task. But we also need to calculate `response_time` on every related action with alert group (ack, resolve, silence). It means that, if we want to calculate in "on-the-fly" we have to do an additional request to db on every ack, silence or resolve action to get alert group logs. 

> Testing
> Prod performance

Not yet, I'll do it later

File: engine/apps/metrics_exporter/tasks.py
Comment (Ferril) [reply engine/apps/metrics_exporter/tasks.py:31]: As far as I understand, celery rate_limit works for all tasks with the same name per worker. I want to rate limit this task for all workers per organization

File: engine/apps/alerts/migrations/0015_alertgroup_response_time.py
Comment (mderynck) [inline engine/apps/alerts/migrations/0015_alertgroup_response_time.py:13]: Currently I don't think we can't handle adding this to the table, migration takes too long in production.  Is there another way we can store it?

File: engine/settings/base.py
Comment (mderynck) [reply engine/settings/base.py:219]: Gateway will not forward traffic to that path (/metrics), see [routes](https://github.com/grafana/backend-enterprise/blob/20c7ca3a215ed54e35b4025bd38145a7df357677/pkg/gateway/routes.go#L727), but we should be sure to test it.

File: engine/apps/metrics_exporter/metrics_cache_manager.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/metrics_cache_manager.py:97]: todo comment

File: engine/apps/metrics_exporter/metrics_collectors.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/metrics_collectors.py:54]: todo comment + print

File: engine/apps/metrics_exporter/metrics_collectors.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/metrics_collectors.py:79]: todo/print

File: engine/apps/metrics_exporter/metrics_collectors.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/metrics_collectors.py:100]: todo/print

File: engine/apps/metrics_exporter/metrics_collectors.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/metrics_collectors.py:111]: todo/print

File: engine/apps/metrics_exporter/helpers.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/helpers.py:117]: todo

File: engine/apps/metrics_exporter/helpers.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/helpers.py:105]: todo

File: engine/apps/metrics_exporter/helpers.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/helpers.py:83]: todo

File: engine/apps/alerts/models/alert_group.py
Comment (joeyorlando) [inline engine/apps/alerts/models/alert_group.py:1]: few remaining todo comments in this file

**Review by joeyorlando — COMMENTED** [review]
it's looking really good! few notes just about remaining todo comments. aside from that some general thoughts to consider:
- unit and/or integration tests. especially for the bulk of your new logic in the celery task, `MetricsCacheManager` and `ApplicationMetricsCollector` classes, added `AlertGroup` model logic, and helper methods
-  logging. consider adding some log statements where you think is appropriate. this will help a lot should we need to debug things later on

File: engine/apps/alerts/models/alert_group.py
Comment (joeyorlando) [inline engine/apps/alerts/models/alert_group.py:1685]: there appears to be 5 other spots in this file's PR diff where you're modifying the object's fields like this. is it possible to deduplicate this into a method on the `AlertGroup` model class?

File: engine/apps/metrics_exporter/metrics_cache_manager.py
Comment (joeyorlando) [inline engine/apps/metrics_exporter/metrics_cache_manager.py:1]: few todo comments in this file

File: engine/apps/alerts/models/alert_group.py
Comment (Ferril) [reply engine/apps/alerts/models/alert_group.py:1685]: I'm not sure how to do it better, since these methods update fields to different values, `fields_to_update` is also different in some places 🤔

File: engine/apps/alerts/models/alert_group.py
Comment (vadimkerr) [inline engine/apps/alerts/models/alert_group.py:1228]: Here it's better to use `alert_groups_to_acknowledge.update` instead of `AlertGroup.all_objects.bulk_update` because regular queryset update is more efficient.

Regular update SQL will look similar to this:

```sql
UPDATE table
SET table.field = 1
WHERE table.id IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
```

Batch update SQL will look similar to this:

```sql
UPDATE table
SET table.field = CASE
                      WHEN (table.id = 1) THEN 1
                      WHEN (table.id = 2) THEN 1
                      WHEN (table.id = 3) THEN 1
                      WHEN (table.id = 4) THEN 1
                      WHEN (table.id = 5) THEN 1
                      WHEN (table.id = 6) THEN 1
                      WHEN (table.id = 7) THEN 1
                      WHEN (table.id = 8) THEN 1
                      WHEN (table.id = 9) THEN 1
                      WHEN (table.id = 10) THEN 1
                      ELSE NULL END
WHERE table.id IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
```

File: engine/apps/alerts/models/alert_group.py
Comment (vadimkerr) [inline engine/apps/alerts/models/alert_group.py:1315]: Use `alert_groups_to_resolve.update`

File: engine/apps/alerts/models/alert_group.py
Comment (vadimkerr) [inline engine/apps/alerts/models/alert_group.py:1458]: Use `alert_groups_to_restart_unresolve.update`

File: engine/apps/alerts/models/alert_group.py
Comment (vadimkerr) [inline engine/apps/alerts/models/alert_group.py:1518]: Use `alert_groups_to_restart_unsilence.update`

File: engine/apps/alerts/models/alert_group.py
Comment (vadimkerr) [inline engine/apps/alerts/models/alert_group.py:1745]:
```suggestion
self.save(update_fields=["acknowledged", "acknowledged_at", "acknowledged_by_user", "acknowledged_by"])
```

File: engine/apps/alerts/models/alert_group.py
Comment (vadimkerr) [inline engine/apps/alerts/models/alert_group.py:1767]:
```suggestion
self.save(update_fields=["resolved", "resolved_at", "resolved_by", "resolved_by_user"])
```

File: engine/apps/alerts/models/alert_group.py
Comment (vadimkerr) [inline engine/apps/alerts/models/alert_group.py:1788]:
```suggestion
        self.save(
            update_fields=["silenced_until", "silenced", "silenced_by_user", "silenced_at", "unsilence_task_uuid"]
        )
```

File: engine/apps/alerts/tasks/unsilence.py
Comment (vadimkerr) [inline engine/apps/alerts/tasks/unsilence.py:37]: Maybe it's better to update metrics in `AlertGroup.un_silence`?

File: engine/apps/metrics_exporter/helpers.py
Comment (vadimkerr) [inline engine/apps/metrics_exporter/helpers.py:26]: Q: Is this here because of circular import?

File: engine/apps/metrics_exporter/helpers.py
Comment (vadimkerr) [inline engine/apps/metrics_exporter/helpers.py:175]: nit: Maybe it's possible to simplify this a bit? It's hard to keep context when there are 4 levels of indentation.

**Review by vadimkerr — CHANGES_REQUESTED** [review]

File: engine/apps/alerts/models/alert_group.py
Comment (matiasb) [reply engine/apps/alerts/models/alert_group.py:520]: The idea is to reconsider adding response_time as an `AlertGroup` field (when we unblock migrations for that), so this is somehow keeping a similar behavior to what we would have if the column is here (we are assuming we would want this in the long-term). If after revisiting this in the short-term we think we want to keep a separated model, it makes sense to move things there.

File: engine/apps/alerts/models/alert_group.py
Comment (matiasb) [reply engine/apps/alerts/models/alert_group.py:529]: Ack, fwiw I like the explicitness of knowing what's triggering the exception (if there is one), but no objections to change it.

File: engine/apps/alerts/models/alert_group.py
Comment (matiasb) [reply engine/apps/alerts/models/alert_group.py:568]: Right, I was thinking something more like a context manager, but was leaving for a next iteration (because I was also thinking we may want to do something around the state transitions related to this).

File: engine/apps/alerts/constants.py
Comment (matiasb) [reply engine/apps/alerts/constants.py:15]: Makes sense

File: engine/apps/alerts/tasks/unsilence.py
Comment (matiasb) [reply engine/apps/alerts/tasks/unsilence.py:37]: Updating metrics inside the state transition methods is tricky because you may lose track of the original state in some cases (e.g. we call un_silence in multiple places, for example we may call this during acknowledge), and also because bulk updates don't go through these methods.

File: engine/apps/metrics_exporter/helpers.py
Comment (matiasb) [reply engine/apps/metrics_exporter/helpers.py:26]: I don't think so, will change

File: engine/apps/metrics_exporter/helpers.py
Comment (matiasb) [reply engine/apps/metrics_exporter/helpers.py:175]: Checking, not seeing an easy refactoring atm, but for sure maybe something to revisit later when we deal with more metrics.

File: engine/apps/alerts/migrations/0016_alertgroupmetrics.py
Comment (joeyorlando) [inline engine/apps/alerts/migrations/0016_alertgroupmetrics.py:18]: what about storing this on the `alert_group` table itself?

File: engine/apps/metrics_exporter/helpers.py
Comment (matiasb) [reply engine/apps/metrics_exporter/helpers.py:26]: Double-checked, there is a circular import indeed (through alerts/alert group), keeping for now.

File: engine/apps/alerts/migrations/0016_alertgroupmetrics.py
Comment (matiasb) [reply engine/apps/alerts/migrations/0016_alertgroupmetrics.py]: :+1:  Switched back to the `alert_group` column implementation.

File: engine/apps/alerts/constants.py
Comment (joeyorlando) [inline engine/apps/alerts/constants.py:19]: I've yet to try this out, but with Python 3.11 there is a new class, [`StrEnum`](https://docs.python.org/3/library/enum.html#enum.StrEnum), might be worth giving a try

**Review by mderynck — APPROVED** [review]

**Review by joeyorlando — APPROVED** [review]

**Pull request merged**
